lib/metrics/context.js

```diff
@@ -6,43 +6,178 @@
-      data.context = metadata.context
-      data.entrypoint = metadata.entrypoint
-      data.migration = metadata.migration
-      data.service = metadata.service
-
-      if (! doNotTrack) {
-        data.utm_campaign = metadata.utmCampaign
-        data.utm_content = metadata.utmContent
-        data.utm_medium = metadata.utmMedium
-        data.utm_source = metadata.utmSource
-        data.utm_term = metadata.utmTerm
+var SCHEMA = isA.object({
+  flowId: isA.string().length(64).regex(HEX).optional(),
+  flowBeginTime: isA.number().integer().positive().optional(),
+  context: isA.string().optional(),
+  entrypoint: isA.string().optional(),
+  migration: isA.string().optional(),
+  service: isA.string().optional(),
+  utmCampaign: isA.string().optional(),
+  utmContent: isA.string().optional(),
+  utmMedium: isA.string().optional(),
+  utmSource: isA.string().optional(),
+  utmTerm: isA.string().optional()
+}).and('flowId', 'flowBeginTime').optional()
+
+var NOP = function () {
+  return P.resolve()
+}
+
+var NULL_MEMCACHED = {
+  delAsync: NOP,
+  getAsync: NOP,
+  setAsync: NOP
+}
+
+module.exports = function (log, config) {
+  var _memcached
+
+  return {
+    schema: SCHEMA,
+    save: save,
+    copy: copy,
+    remove: remove
+  }
+
+  /**
+   * Saves metrics context metadata against a token.
+   *
+   * @param token    token to save the metadata against
+   * @param events   array of event names that constitute a flow
+   * @param metadata metrics context metadata
+   */
+  function save (token, events, metadata) {
+    if (! token || ! events || ! metadata) {
+      return P.resolve()
+    }
+
+    if (typeof events === 'string') {
+      events = [ events ]
+    }
+
+    if (! Array.isArray(events)) {
+      return P.resolve()
+    }
+
+    var memcached = getMemcached()
+
+    return P.all(events.map(function (event) {
+      return memcached.setAsync(getKey(token, event), metadata, config.memcached.lifetime)
+        .catch(function (err) {
+          log.error({ op: 'metricsContext.save', err: err })
+        })
+    }))
+  }
+
+  function getMemcached () {
+    try {
+      if (config.memcached.address === 'none') {
+        return NULL_MEMCACHED
       }
+
+      if (! _memcached) {
+        _memcached = new Memcached(config.memcached.address, {
+          timeout: 500,
+          retries: 1,
+          retry: 1000,
+          reconnect: 1000,
+          idle: config.memcached.idle,
+          namespace: 'fxa-metrics~'
+        })
+      }
+
+      return _memcached
+    } catch (err) {
+      log.error({ op: 'metricsContext.getMemcached', err: err })
+
+      return NULL_MEMCACHED
     }
+  }
+
+  /**
+   * Copies metrics context metadata to data, using either metadata
+   * passed in from a request or previously-saved metadata for a
+   * token.
+   *
+   * @param data    target object
+   * @param request request object
+   * @param event   event name
```

Comment:
docs nit: this has a meaningful return value so it should probably be documented as well